### DISCUSSION TITLE
How to create your own Wechaty Puppet Service Token with the Web Protocol

### DISCUSSION THREAD

**Discussion by huan** [wechaty/wechaty#1986] [Announcements]
# Update Mar 29, 2021

We have updated our official documentation at <https://wechaty.js.org/docs/puppet-services/diy> for how to create a token gateway server by yourself.

# ⬆️ Please read the above documentation before you continue reading the following information

Since we have our branding new Multi-lang Wechaty ready to use from this week, it would be great for our Python, Java, and Go developers to have a try with them!

1. https://github.com/wechaty/python-wechaty-getting-started
1. https://github.com/wechaty/java-wechaty-getting-started
1. https://github.com/wechaty/go-wechaty-getting-started

All of the above gettings started repos will be able to run out-of-the-box. However, they will all need a TOKEN for [wechaty-puppet-hostie](https://github.com/wechaty/wechaty-puppet-hostie).

**How to get this Hostie Token?**

1. If you have token before (like a padplus token), you can go to #1985; else
1. If you have no token before, this post will teach you a very easy way to **Create One Wechaty Hostie Token** by yourself.

## Requirements

A server with:

1. Public IP
1. Public Port
1. Docker

That's all you need!

### 1 Pull the latest Wechaty Docker Image

```sh
docker pull wechaty/wechaty
```

### 2 Select a random string as your `wechaty-puppet-hostie` token

```sh
# Set the `wechaty-puppet-hostie` token for your hostie
export WECHATY_TOKEN=puppet_hostie_your_token
```

### 3 Start your Wechaty Puppet Service Provider Server:

Notice: 

1. The server IP must be public on the internet, and 
1. the `WECHATY_HOSTIE_PORT` must be accessible on the internet.

```sh
# Set port for your hostie service: must be published accessible on the internet
export WECHATY_HOSTIE_PORT=8788

export WECHATY_PUPPET=wechaty-puppet-puppeteer
export WECHATY_LOG=verbose

docker run \
  --rm \
  -ti \
  -e WECHATY_LOG \
  -e WECHATY_PUPPET \
  -e WECHATY_HOSTIE_PORT \
  -e WECHATY_TOKEN \
  -p "$WECHATY_HOSTIE_PORT" \
  wechaty/wechaty
```

### 4 That's it, you are set!

Your token is ready to use by Python, Java, and Go Wechaty!

Link to: #1927

## Use Your Wechaty Token

From any computer, you can run the following command to use your token:

```sh
export WECHATY_LOG=verbose 
export WECHATY_PUPPET=wechaty-puppet-hostie
export WECHATY_PUPPET_HOSTIE_TOKEN=puppet_hostie_your_token

# Run your Wechaty bot
node your_wechaty_bot.js
```

### 1 How to confirm that my hostie token is working?

Answer: Visit our RESTFUL API at <https://api.chatie.io/v0/hosties/>__TOKEN__, where `TOKEN` is your token:

1. If you see an IP with '0.0.0.0' then it's invalid.
1. If you can get a meaningful IP/PORT from this API, then your TOKEN is valid.

The following is an example for INVALID token:

```json
{"ip":"0.0.0.0","port":0}
```

Please check your token before you using it with Python, Java, and Go Wechaty.

## Update: ~~`zixia/wechaty`~~ has been renamed to `wechaty/wechaty`

Please use the `wechaty/wechaty` docker image name after Aug 2020.

## Resources

The toke is provided from our Wechaty community **Puppet Service Providers**.

Learn more about how to get a token from the Puppet Service Providers at our website: [Wechaty Puppet Services](https://wechaty.js.org/docs/puppet-services)

**Comment by huan** [comment]
@wechaty/python @wechaty/java @wechaty/go @wechaty/scala Please take some time to read this post so that we can all understand how to build a token by ourselves.

Thank you very much!

**Comment by Chaogan-Yan** [comment]
Thanks a lot for your info!
I am trying to build Token on my own server. However, I got the error message below when starting docker. Is this related to my own server ports? Or how should I check and solve such a problem? Thanks!

07:39:56 VERB Io initWebSocket()
07:40:24 WARN Io initWebSocket() error event[getaddrinfo EAI_AGAIN api.chatie.io]
07:40:24 VERB Io send() without a connected websocket, eventBuffer.length = 1
07:40:24 WARN Io initWebSocket() close event[1006: ]
07:40:24 VERB Io reconnect()

**Comment by z1045** [comment]
Hello Huan,

Thanks a lot for the instructions here. Does your program support puppet_padplus instead of puppet_puppeteer since web interface has been effectively closed by Tencent?

Zhao

**Comment by huan** [comment]
@z1045 You are welcome.

Of course our program support puppet_padplus (and all other protocols). Please refer to #1985 as I mentioned in the post.

> If you have token before (like a padplus token), you can go to #1985

**Comment by wordgao** [comment]
都是中国人，为何要用英语呢，看的有点头疼，也不是很理解上面说的问题。

**Comment by ZakaryCode** [comment]
Is there a chance that BOT and service can run in the same container like this.

```
export WECHATY_HOSTIE_PORT=8788

export WECHATY_PUPPET_HOSTIE_TOKEN = wechaty-token
export WECHATY_PUPPET=wechaty-puppet-hostie
export WECHATY_LOG=verbose

docker run \
  --rm \
  -ti \
  -e WECHATY_LOG \
  -e WECHATY_PUPPET \
  -e WECHATY_HOSTIE_PORT \
  -e WECHATY_TOKEN=WECHATY_PUPPET_HOSTIE_TOKEN \
  -e WECHATY_PUPPET_HOSTIE_TOKEN \
  -p "$WECHATY_HOSTIE_PORT" \
  wechaty/wechaty
```

**Comment by hjd31** [comment]
我用https://api.chatie.io/v0/hosties/3f37e1a5-ca4f-4f61-b6ba-25e8a019d1c1测试了我的令牌，得到了 json 数据：
{"host":"1.15.51.67","ip":"1.15.51.67","port":15023}
然后我在Java中运行时，报错了
start() rejection:
io.grpc.StatusRuntimeException: UNAUTHENTICATED: Invalid Wechaty TOKEN "1.15.51.67:15023"
	at io.grpc.stub.ClientCalls.toStatusRuntimeException(ClientCalls.java:244) ~[grpc-stub-1.29.0.jar:1.29.0]
	at io.grpc.stub.ClientCalls.getUnchecked(ClientCalls.java:225) ~[grpc-stub-1.29.0.jar:1.29.0]
	at io.grpc.stub.ClientCalls.blockingUnaryCall(ClientCalls.java:142) ~[grpc-stub-1.29.0.jar:1.29.0]
	at io.github.wechaty.grpc.PuppetGrpc$PuppetBlockingStub.start(PuppetGrpc.java:2744) ~[grpc-0.16.1.jar:?]
	at io.github.wechaty.grpc.GrpcPuppet.start(GrpcPuppet.kt:78) [wechaty-puppet-hostie-0.1.5-SNAPSHOT.jar:?]
	at io.github.wechaty.Wechaty.start(Wechaty.kt:52) [wechaty-0.1.5-SNAPSHOT.jar:?]
	at io.github.wechaty.DingDongBot.main(DingDongBot.java:25) [classes/:?]

代码如下：
Wechaty bot = Wechaty.instance("3f37e1a5-ca4f-4f61-b6ba-25e8a019d1c1")
                .onScan((qrcode, statusScanStatus, data) -> System.out.println(QrcodeUtils.getQr(qrcode)))
                .onLogin(user -> System.out.println(user))
                .onMessage(message -> {
                    Room room = message.room();
                    String text = message.text();
                    if (StringUtils.equals(text, "#ding")) {
                        if (room != null) {
                            room.say("dong");
                        }
                    }
                }).start(true);

**Reply by sumnhao** [reply]
解决了吗

**Reply by heyiqian0831** [reply]
start() rejection:
java.net.SocketTimeoutException: Read timed out我是这个，大佬你解决了吗
